src/Equinox.Core/Cache.fs

```diff
@@ -41,34 +44,87 @@ module private CachingStrategy =
+        | Ok validatedCachedState -> return validatedCachedState
+        | Error (current, newInstance) ->
+            let _ = System.Threading.Interlocked.CompareExchange(&cell, newInstance, current)
+            let! struct (token, state) as res = cell.Await()
+            if obj.ReferenceEquals(cell, current) then x.MergeUpdates(compare, token, state, timestamp)
+            return res }
 
 type Cache private (inner: MemoryCache) =
+    let tryLoad key =
+        match inner.Get key with
+        | null -> ValueNone
+        | :? CacheEntry<'state> as existingEntry -> existingEntry.TryGetValue()
+        | x -> failwith $"tryLoad Incompatible cache entry %A{x}"
+    let addOrGet key options entry =
+        match inner.AddOrGetExisting(key, entry, CachingStrategy.toPolicy options) with
+        | null -> Ok entry
+        | :? CacheEntry<'state> as existingEntry -> Error existingEntry
+        | x -> failwith $"addOrGet Incompatible cache entry %A{x}"
+    let getElseAddEmptyEntry key options =
+        match addOrGet key options (CacheEntry<'state>.CreateEmpty()) with
+        | Ok fresh -> fresh
+        | Error existingEntry -> existingEntry
+    let addOrMergeCacheEntry compare key options struct (token, state) =
+        let timestamp = System.Diagnostics.Stopwatch.GetTimestamp()
+        let entry = CacheEntry(token, state, timestamp)
+        match addOrGet key options entry with
+        | Ok _ -> () // Our fresh one got added
+        | Error existingEntry -> existingEntry.MergeUpdates(compare, token, state, timestamp)
     new (name, sizeMb: int) =
         let config = System.Collections.Specialized.NameValueCollection(1)
         config.Add("cacheMemoryLimitMegabytes", string sizeMb);
         Cache(new MemoryCache(name, config))
     interface ICache with
-        member _.TryGet key =
-            match inner.Get key with
-            | null -> ValueNone |> Task.FromResult
-            | :? CacheEntry<'state> as existingEntry -> ValueSome existingEntry.Value |> Task.FromResult
-            | x -> failwithf "TryGet Incompatible cache entry %A" x
-        member _.UpdateIfNewer(key, compare, options, token, state) =
-            let freshEntry = CacheEntry(token, state)
-            match inner.AddOrGetExisting(key, freshEntry, CachingStrategy.toPolicy options) with
-            | null -> Task.FromResult()
-            | :? CacheEntry<'state> as existingEntry -> existingEntry.UpdateIfNewer(compare, freshEntry); Task.FromResult()
-            | x -> failwithf "UpdateIfNewer Incompatible cache entry %A" x
+        // if there's a non-zero maxStaleness, concurrent read attempts share the roundtrip (and its fate, if it throws)
+        member x.Load(key, maxStaleness, compare, options, loadOrReload) = task {
+            let loadOrReload maybeBaseState () =
+                let act = System.Diagnostics.Activity.Current
+                if act <> null then act.AddCacheHit(ValueOption.isSome maybeBaseState) |> ignore
```

Comment:
🙏

Reply:
BTW not certain whether there's value in separated metrics for incremental loads due to caching vs AllowStale roundtrips saved, or whether a cache hit is the main point - will be pondering